atproto/auth/permission.go

```diff
@@ -0,0 +1,288 @@
+		scope = scope + "?" + params.Encode()
+	}
+	return scope
+}
+
+// Parses a permission scope string (as would be found as a component of an OAuth scope string) into a [Permission].
+//
+// This function is strict: it is case sensitive, verifies field syntax, and will throw an error on unknown parameters/fields. Note that calling code is usually supposed to simply skip any permission which cause such errors, not reject entire requests.
+func ParsePermissionString(scope string) (*Permission, error) {
+	g, err := ParseGenericScope(scope)
+	if err != nil {
+		return nil, err
+	}
+
+	p := Permission{
+		Type:     "permission",
+		Resource: g.Resource,
+	}
+
+	switch g.Resource {
+	case "account":
+		for k, _ := range g.Params {
+			if !(k == "attr" || k == "action") {
+				return nil, fmt.Errorf("%w: unsupported 'account' param: %s", ErrInvalidPermissionParams, k)
+			}
+		}
+		if g.Params.Has("attr") {
+			if g.Positional != "" || len(g.Params["attr"]) != 1 {
+				return nil, ErrInvalidPermissionParams
+			}
+			p.Attribute = g.Params.Get("attr")
+		}
+		if g.Positional != "" {
+			p.Attribute = g.Positional
+		}
+		if p.Attribute == "" {
+			return nil, ErrInvalidPermissionParams
+		}
+		if p.Attribute != "" && p.Attribute != "email" && p.Attribute != "repo" {
+			return nil, ErrInvalidPermissionParams
+		}
+		// TODO: maybe this should not be limited to a single "action" string?
+		if len(g.Params["action"]) > 1 {
+			return nil, ErrInvalidPermissionParams
+		}
+		p.Action = g.Params["action"]
+		for _, act := range p.Action {
+			if act != "read" && act != "manage" {
+				return nil, ErrInvalidPermissionParams
+			}
+		}
+	case "blob":
+		for k, _ := range g.Params {
+			if !(k == "accept") {
+				return nil, fmt.Errorf("%w: unsupported 'blob' param: %s", ErrInvalidPermissionParams, k)
+			}
+		}
+		if g.Params.Has("accept") {
+			if g.Positional != "" {
+				return nil, ErrInvalidPermissionParams
+			}
+			p.Accept = g.Params["accept"]
+		}
+		if g.Positional != "" {
+			p.Accept = []string{g.Positional}
+		}
+		if len(p.Accept) == 0 {
+			return nil, ErrInvalidPermissionParams
+		}
+		for _, acc := range p.Accept {
+			if !validBlobAccept(acc) {
+				return nil, ErrInvalidPermissionParams
+			}
+		}
+	case "identity":
+		for k, _ := range g.Params {
+			if !(k == "attr") {
+				return nil, fmt.Errorf("%w: unsupported 'identity' param: %s", ErrInvalidPermissionParams, k)
+			}
+		}
+		if g.Params.Has("attr") {
+			if g.Positional != "" || len(g.Params["attr"]) != 1 {
+				return nil, ErrInvalidPermissionParams
+			}
+			p.Attribute = g.Params.Get("attr")
+		}
+		if g.Positional != "" {
+			p.Attribute = g.Positional
+		}
+		if p.Attribute != "*" && p.Attribute != "handle" {
+			return nil, ErrInvalidPermissionParams
+		}
+	case "include":
+		for k, _ := range g.Params {
+			if !(k == "nsid" || k == "aud") {
+				return nil, fmt.Errorf("%w: unsupported 'include' param: %s", ErrInvalidPermissionParams, k)
+			}
+		}
+		if g.Params.Has("nsid") {
```

Comment:
I'm not fluent in go: do we check that the positional and name attribute are not set together?